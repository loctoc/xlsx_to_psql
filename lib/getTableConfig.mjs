import XLSX from 'xlsx';
import fs from 'fs';

export function getTableConfigOverrides(tableConfigFile) {
    if (tableConfigFile && fs.existsSync(tableConfigFile)) {
        const tableConfig = JSON.parse(fs.readFileSync(tableConfigFile, 'utf-8'));
        return tableConfig
    }
    return {}
}

export function getTableConfigForAWorkSheet(worksheet, tableConfigFile) {
    const tableConfigOverrides = getTableConfigOverrides(tableConfigFile);
    const range = XLSX.utils.decode_range(worksheet['!ref'] || 'A1');

    // First get all headers and create base configs
    const tableConfig = Array(range.e.c + 1).fill(1).map((_, idx) => {
        const cellAddress = XLSX.utils.encode_cell({ r: 0, c: idx }); // headers
        const cell = worksheet[cellAddress];
        const colVal = cell?.v || `Column${idx + 1}`;
        const colOverrides = tableConfigOverrides?.[colVal] ?? {};
        
        return {
            "header": colVal,
            "sqlColumn": colVal,
            "fieldType": "string",
            "primary": false,
            "notNull": false,
            "skip": false,
            "needIndex": false,
            "isHyperlink": true,
            ...colOverrides
        };
    });

    // Count occurrences of each column name
    const columnNameCounts = {};
    tableConfig.forEach((config) => {
        if (!config.skip) {
            const baseName = config.sqlColumn;
            columnNameCounts[baseName] = (columnNameCounts[baseName] || 0) + 1;
        }
    });

    // Add suffix to all columns that have duplicates
    const seenColumns = {};
    tableConfig.forEach((config) => {
        if (!config.skip) {
            const baseName = config.sqlColumn;
            if (columnNameCounts[baseName] > 1) {
                seenColumns[baseName] = (seenColumns[baseName] || 0) + 1;
                config.sqlColumn = `${baseName}_${seenColumns[baseName]}`;
            }
        }
    });

    return tableConfig;
}