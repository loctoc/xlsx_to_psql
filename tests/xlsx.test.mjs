import { describe, test, expect } from 'vitest';
import XLSX from 'xlsx';
import path from 'path';
import { fileURLToPath } from 'url';
import { getTableConfigForAWorkSheet } from '../lib/getTableConfig.mjs';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

describe('parseAndTransformXLSX', () => {
  test('should parse and transform XLSX data', () => {
    // TODO: Implement test
  })
});

describe('getTableConfigForAWorkSheet', () => {
  test('should handle duplicate column names correctly', () => {
    // Load the test Excel file
    const workbook = XLSX.readFile(path.join(__dirname, 'duplicae-columns-test.xlsx'));
    const worksheet = workbook.Sheets[workbook.SheetNames[0]];

    // Get table config
    const tableConfig = getTableConfigForAWorkSheet(worksheet, null);

    // Expected column names after handling duplicates
    const expectedColumns = [
      'Form Name',
      'Submission Number',
      'Sent At',
      'Sender Name',
      'Sender Identifier',
      'Sender Department',
      'Sender Designation',
      'Sender Location',
      'Sender Division',
      'Sender Sub Division',
      'Current Status',
      'Remarks',
      'Product1',
      'Evidence_1',
      'Product2',
      'Evidence_2',
      'Product3',
      'Evidence_3'
    ];

    // Test that we have the correct number of columns
    expect(tableConfig.length).toBe(expectedColumns.length);

    // Test that each column has the expected name
    tableConfig.forEach((config, index) => {
      expect(config.sqlColumn).toBe(expectedColumns[index]);
    });

    // Specifically test the duplicate 'Evidence' columns
    const evidenceColumns = tableConfig.filter(config => 
      config.header === 'Evidence'
    );
    
    expect(evidenceColumns.length).toBe(3);
    expect(evidenceColumns[0].sqlColumn).toBe('Evidence_1');
    expect(evidenceColumns[1].sqlColumn).toBe('Evidence_2');
    expect(evidenceColumns[2].sqlColumn).toBe('Evidence_3');
    expect(tableConfig).toMatchSnapshot();
  });
  test.skip('Critical_Machine_Checks_Form_responses.xlsx real test case 1', () => {
    // Load the test Excel file
    const workbook = XLSX.readFile(path.join(__dirname, 'Real_TestCase_1.xlsx'));
    const worksheet = workbook.Sheets[workbook.SheetNames[0]];
    const tableConfig = getTableConfigForAWorkSheet(worksheet, path.join(__dirname, 'table-config.json'));
    expect(tableConfig).toMatchSnapshot();
  })
});