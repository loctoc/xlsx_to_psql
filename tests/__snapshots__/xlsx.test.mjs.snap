// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getTableConfigForAWorkSheet > Critical_Machine_Checks_Form_responses.xlsx real test case 1 1`] = `
[
  {
    "fieldType": "string",
    "header": "Form Name",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Form Name",
  },
  {
    "fieldType": "string",
    "header": "Submission Number",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Submission Number",
  },
  {
    "fieldType": "timestamp",
    "header": "Sent At",
    "isHyperlink": true,
    "needIndex": true,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Sent At",
  },
  {
    "fieldType": "string",
    "header": "Sender Name",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Sender Name",
  },
  {
    "fieldType": "string",
    "header": "Sender Identifier",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Sender Identifier",
  },
  {
    "fieldType": "string",
    "header": "Sender Department",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Sender Department",
  },
  {
    "fieldType": "string",
    "header": "Sender Designation",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Sender Designation",
  },
  {
    "fieldType": "string",
    "header": "Sender Location",
    "isHyperlink": true,
    "needIndex": true,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Sender Location",
  },
  {
    "fieldType": "string",
    "header": "Sender Division",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Sender Division",
  },
  {
    "fieldType": "string",
    "header": "Sender Sub Division",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Sender Sub Division",
  },
  {
    "fieldType": "string",
    "header": "Current Status",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Current Status",
  },
  {
    "fieldType": "string",
    "header": "Remarks",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Remarks",
  },
  {
    "fieldType": "string",
    "header": "Name",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Name",
  },
  {
    "fieldType": "string",
    "header": "Response Link",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Response Link",
  },
  {
    "fieldType": "string",
    "header": "Stage 1: Date",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Stage 1: Date",
  },
  {
    "fieldType": "string",
    "header": "Stage 1: Name",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Stage 1: Name",
  },
  {
    "fieldType": "string",
    "header": "Stage 1: Action",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Stage 1: Action",
  },
  {
    "fieldType": "string",
    "header": "Stage 1: Location",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Stage 1: Location",
  },
  {
    "fieldType": "string",
    "header": "Select your location",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Select your location",
  },
  {
    "fieldType": "string",
    "header": "ROBO COUPE MACHINE - GRATER",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "ROBO COUPE MACHINE - GRATER",
  },
  {
    "fieldType": "string",
    "header": "Attached photo",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Attached photo",
  },
  {
    "fieldType": "string",
    "header": "Dough mixer (big)",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Dough mixer (big)",
  },
  {
    "fieldType": "string",
    "header": "Attach picture of machine blade / whisk etc",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Attach picture of machine blade / whisk etc_1",
  },
  {
    "fieldType": "string",
    "header": "Coffee grinder",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Coffee grinder",
  },
  {
    "fieldType": "string",
    "header": "Attach picture of machine blade / whisk etc",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Attach picture of machine blade / whisk etc_2",
  },
  {
    "fieldType": "string",
    "header": "Mixer 1 (small)",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Mixer 1 (small)",
  },
  {
    "fieldType": "string",
    "header": "Attach picture of machine blade / whisk etc",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Attach picture of machine blade / whisk etc_3",
  },
  {
    "fieldType": "string",
    "header": "Mixer 2 (small)",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Mixer 2 (small)",
  },
  {
    "fieldType": "string",
    "header": "Attach picture of machine blade / whisk etc",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Attach picture of machine blade / whisk etc_4",
  },
  {
    "fieldType": "string",
    "header": "Blender 1",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Blender 1",
  },
  {
    "fieldType": "string",
    "header": "Attach picture of machine blade / whisk etc",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Attach picture of machine blade / whisk etc_5",
  },
  {
    "fieldType": "string",
    "header": "Blender 2",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Blender 2",
  },
  {
    "fieldType": "string",
    "header": "Attach picture of machine blade / whisk etc",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Attach picture of machine blade / whisk etc_6",
  },
  {
    "fieldType": "string",
    "header": "Robot coupe R 301",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Robot coupe R 301",
  },
  {
    "fieldType": "string",
    "header": "Attach picture of machine blade / whisk etc",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Attach picture of machine blade / whisk etc_7",
  },
  {
    "fieldType": "string",
    "header": "Slicer 1",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Slicer 1",
  },
  {
    "fieldType": "string",
    "header": "Attach picture of machine blade / whisk etc",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Attach picture of machine blade / whisk etc_8",
  },
  {
    "fieldType": "string",
    "header": "Dough sheeter",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Dough sheeter",
  },
  {
    "fieldType": "string",
    "header": "Attach picture of machine blade / whisk etc",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Attach picture of machine blade / whisk etc_9",
  },
  {
    "fieldType": "string",
    "header": "Mincing machine (big)",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Mincing machine (big)",
  },
  {
    "fieldType": "string",
    "header": "Attach picture of machine blade / whisk etc",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Attach picture of machine blade / whisk etc_10",
  },
  {
    "fieldType": "string",
    "header": "Mincer machine (small)",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Mincer machine (small)",
  },
  {
    "fieldType": "string",
    "header": "Attach picture of machine blade / whisk etc",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Attach picture of machine blade / whisk etc_11",
  },
  {
    "fieldType": "string",
    "header": "Robot coupe R 10",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Robot coupe R 10",
  },
  {
    "fieldType": "string",
    "header": "Attach picture of machine blade",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Attach picture of machine blade_1",
  },
  {
    "fieldType": "string",
    "header": "Hummus machine R 45",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Hummus machine R 45",
  },
  {
    "fieldType": "string",
    "header": "Attach picture of machine blade",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Attach picture of machine blade_2",
  },
  {
    "fieldType": "string",
    "header": "Hummus machine R 60",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Hummus machine R 60",
  },
  {
    "fieldType": "string",
    "header": "Attach picture of machine blade",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Attach picture of machine blade_3",
  },
  {
    "fieldType": "string",
    "header": "Robot coupe CL 50",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Robot coupe CL 50",
  },
  {
    "fieldType": "string",
    "header": "Attach picture of machine blade",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Attach picture of machine blade_4",
  },
  {
    "fieldType": "string",
    "header": "Tin / Can Cutter",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Tin / Can Cutter",
  },
  {
    "fieldType": "string",
    "header": "Attach picture of machine blade",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Attach picture of machine blade_5",
  },
  {
    "fieldType": "string",
    "header": "Knives",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Knives",
  },
  {
    "fieldType": "string",
    "header": "Attach picture of machine blade",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Attach picture of machine blade_6",
  },
  {
    "fieldType": "string",
    "header": "Carrot juice machine",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Carrot juice machine",
  },
  {
    "fieldType": "string",
    "header": "Attach picture of machine blade",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Attach picture of machine blade_7",
  },
  {
    "fieldType": "string",
    "header": "Other Machine not in list above",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Other Machine not in list above",
  },
  {
    "fieldType": "string",
    "header": "Checklist completed by",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Checklist completed by",
  },
]
`;

exports[`getTableConfigForAWorkSheet > should handle duplicate column names correctly 1`] = `
[
  {
    "fieldType": "string",
    "header": "Form Name",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Form Name",
  },
  {
    "fieldType": "string",
    "header": "Submission Number",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Submission Number",
  },
  {
    "fieldType": "string",
    "header": "Sent At",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Sent At",
  },
  {
    "fieldType": "string",
    "header": "Sender Name",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Sender Name",
  },
  {
    "fieldType": "string",
    "header": "Sender Identifier",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Sender Identifier",
  },
  {
    "fieldType": "string",
    "header": "Sender Department",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Sender Department",
  },
  {
    "fieldType": "string",
    "header": "Sender Designation",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Sender Designation",
  },
  {
    "fieldType": "string",
    "header": "Sender Location",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Sender Location",
  },
  {
    "fieldType": "string",
    "header": "Sender Division",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Sender Division",
  },
  {
    "fieldType": "string",
    "header": "Sender Sub Division",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Sender Sub Division",
  },
  {
    "fieldType": "string",
    "header": "Current Status",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Current Status",
  },
  {
    "fieldType": "string",
    "header": "Remarks",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Remarks",
  },
  {
    "fieldType": "string",
    "header": "Product1",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Product1",
  },
  {
    "fieldType": "string",
    "header": "Evidence",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Evidence_1",
  },
  {
    "fieldType": "string",
    "header": "Product2",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Product2",
  },
  {
    "fieldType": "string",
    "header": "Evidence",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Evidence_2",
  },
  {
    "fieldType": "string",
    "header": "Product3",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Product3",
  },
  {
    "fieldType": "string",
    "header": "Evidence",
    "isHyperlink": true,
    "needIndex": false,
    "notNull": false,
    "primary": false,
    "skip": false,
    "sqlColumn": "Evidence_3",
  },
]
`;
